File: api/src/schema/billing.rs
use async_graphql::SimpleObject;

use super::*;
use crate::billing::{self, StripeConfig};

/// Billing interval for a Crew subscription.
#[derive(async_graphql::Enum, Copy, Clone, Eq, PartialEq, Debug)]
pub enum BillingInterval {
    Monthly,
    Annual,
}

/// The org's billing posture for the client: plan, status, usage, and limits
/// (`-1` = unlimited). Drives the upgrade gate, caps, and the Billing settings page.
#[derive(SimpleObject)]
pub struct BillingInfo {
    /// "solo" (free) or "crew" (paid).
    pub plan: String,
    pub status: Option<String>,
    pub current_period_end: Option<chrono::DateTime<chrono::Utc>>,
    pub cancel_at_period_end: bool,
    /// Read-only lock (lapsed sub over the Solo caps).
    pub restricted: bool,
    pub can_export: bool,
    pub projects: i64,
    pub admins: i64,
    pub non_admin: i64,
    pub max_projects: i64,
    pub max_admins: i64,
    pub max_non_admin: i64,
    /// Emails of this org's admins, so non-admin members can reach out about
    /// upgrading/managing the subscription.
    pub admin_emails: Vec<String>,
}

#[derive(Default)]
pub struct BillingQuery;

#[Object]
impl BillingQuery {
    /// The caller org's plan, subscription status, usage, and limits.
    async fn billing(&self, ctx: &Context<'_>) -> Result<BillingInfo> {
        let auth = require_auth(ctx)?;
        let pool = pool(ctx)?;
        let mut b = billing::org_billing(pool, auth.org_id).await?;

        // Self-heal: a paid org with no renewal date means a webhook was missed or
        // didn't carry the period (e.g. a Stripe schema change). Pull the live
        // subscription state from Stripe once; subsequent loads skip this.
        if b.paid() && b.current_period_end.is_none() {
            if let Ok(cfg) = ctx.data::<StripeConfig>() {
                if cfg.enabled() {
                    let sub: Option<(Option<String>,)> =
                        sqlx::query_as("SELECT stripe_subscription_id FROM orgs WHERE id = $1")
                            .bind(auth.org_id)
                            .fetch_optional(pool)
                            .await?;
                    if let Some((Some(sub_id),)) = sub {
                        let _ = billing::resync_subscription(cfg, pool, &sub_id).await;
                        b = billing::org_billing(pool, auth.org_id).await?;
                    }
                }
            }
        }

        let admin_emails: Vec<String> = sqlx::query_scalar(
            "SELECT email FROM users WHERE org_id = $1 AND role = 'admin' ORDER BY email",
        )
        .bind(auth.org_id)
        .fetch_all(pool)
        .await?;
        // Caps come from the plan catalog: Crew resolves to unlimited (-1).
        let plan = b.plan();
        let limits = plan.limits();
        Ok(BillingInfo {
            plan: plan.as_str().to_string(),
            status: b.status.clone(),
            current_period_end: b.current_period_end,
            cancel_at_period_end: b.cancel_at_period_end,
            restricted: b.restricted(),
            can_export: b.can_export(),
            projects: b.projects,
            admins: b.admins,
            non_admin: b.non_admin,
            max_projects: limits.projects,
            max_admins: limits.admins,
            max_non_admin: limits.non_admin,
            admin_emails,
        })
    }
}

/// A gated feature in the plan catalog (static metadata; drives the web's upgrade
/// dialogs + selling points). See `crate::plan`.
#[derive(SimpleObject)]
pub struct PlanFeature {
    /// Stable snake_case id (upgrade dialogs key off this).
    pub key: String,
    pub label: String,
    pub blurb: String,
    pub min_plan: crate::plan::Plan,
}

/// A plan's usage caps (`-1` = unlimited).
#[derive(SimpleObject)]
pub struct PlanLimits {
    pub plan: crate::plan::Plan,
    pub max_projects: i64,
    pub max_admins: i64,
    pub max_non_admin: i64,
}

/// The static plan → capability catalog: the single source of truth for what each
/// plan unlocks. The web renders upgrade UI + selling points from this instead of
/// hard-coding feature lists.
#[derive(SimpleObject)]
pub struct PlanCatalog {
    pub features: Vec<PlanFeature>,
    pub plans: Vec<PlanLimits>,
}

#[derive(Default)]
pub struct PlanCatalogQuery;

#[Object]
impl PlanCatalogQuery {
    /// The static plan/feature catalog. No auth required — it's marketing/gating
    /// metadata, safe to render on public pricing surfaces.
    async fn plan_catalog(&self, _ctx: &Context<'_>) -> Result<PlanCatalog> {
        use crate::plan::{Feature, Plan};
        let features = Feature::all()
            .iter()
            .map(|&f| {
                let m = f.meta();
                PlanFeature {
                    key: m.key.to_string(),
                    label: m.label.to_string(),
                    blurb: m.blurb.to_string(),
                    min_plan: m.min_plan,
                }
            })
            .collect();
        let plans = Plan::all()
            .iter()
            .map(|&p| {
                let l = p.limits();
                PlanLimits {
                    plan: p,
                    max_projects: l.projects,
                    max_admins: l.admins,
                    max_non_admin: l.non_admin,
                }
            })
            .collect();
        Ok(PlanCatalog { features, plans })
    }
}

#[derive(Default)]
pub struct BillingMutation;

#[Object]
impl BillingMutation {
    /// Starts a hosted Stripe Checkout for the Crew plan; returns the redirect URL
    /// the client sends the browser to. Admin-only.
    async fn create_checkout_session(
        &self,
        ctx: &Context<'_>,
        interval: BillingInterval,
    ) -> Result<String> {
        let auth = require_admin(ctx)?;
        let pool = pool(ctx)?;
        let cfg = ctx.data::<StripeConfig>()?;
        if !cfg.enabled() {
            return Err(async_graphql::Error::new("billing is not configured"));
        }
        let (email,): (String,) = sqlx::query_as("SELECT email FROM users WHERE id = $1")
            .bind(auth.user_id)
            .fetch_one(pool)
            .await?;
        let customer = billing::ensure_customer(cfg, pool, auth.org_id, &email)
            .await
            .map_err(async_graphql::Error::new)?;
        let price = match interval {
            BillingInterval::Monthly => &cfg.price_monthly,
            BillingInterval::Annual => &cfg.price_annual,
        };
        if price.is_empty() {
            return Err(async_graphql::Error::new("billing price is not configured"));
        }
        billing::create_checkout_session(cfg, &customer, price, auth.org_id)
            .await
            .map_err(async_graphql::Error::new)
    }

    /// Opens the Stripe Customer Portal (manage card / cancel / invoices); returns
    /// the redirect URL. Admin-only.
    async fn create_billing_portal_session(&self, ctx: &Context<'_>) -> Result<String> {
        let auth = require_admin(ctx)?;
        let pool = pool(ctx)?;
        let cfg = ctx.data::<StripeConfig>()?;
        if !cfg.enabled() {
            return Err(async_graphql::Error::new("billing is not configured"));
        }
        let (customer,): (Option<String>,) =
            sqlx::query_as("SELECT stripe_customer_id FROM orgs WHERE id = $1")
                .bind(auth.org_id)
                .fetch_one(pool)
                .await?;
        let customer = customer
            .ok_or_else(|| async_graphql::Error::new("no Stripe customer for this organization"))?;
        billing::create_portal_session(cfg, &customer)
            .await
            .map_err(async_graphql::Error::new)
    }
}

File: web/src/lib/billing.ts
'use client';

import { useCallback, useEffect, useState } from 'react';
import { toast } from 'sonner';

import { graphql } from '@/lib/gql';
import { gql } from '@/lib/graphql';
import { PLAN_CATALOG_QUERY, type PlanCatalog } from '@/lib/plan';
import { errMsg } from '@/lib/utils';

// The plan catalog (types, query, PRICING, and pure helpers) lives in the neutral
// `@/lib/plan` module so server components can use it too. Re-exported here so the
// existing `@/lib/billing` import sites keep working.
export {
  crewSellingPoints,
  featureMeta,
  fetchPlanCatalog,
  PLAN_CATALOG_QUERY,
  type PlanCatalog,
  type PlanFeature,
  type PlanLimits,
  PRICING,
} from '@/lib/plan';

/** The org's billing posture, mirroring the API's `BillingInfo`. Limits use
 *  `-1` for "unlimited" (Crew). `plan` is `'solo'` (free) or `'crew'` (paid). */
export interface Billing {
  adminEmails: string[];
  admins: number;
  cancelAtPeriodEnd: boolean;
  canExport: boolean;
  currentPeriodEnd: string | null;
  maxAdmins: number;
  maxNonAdmin: number;
  maxProjects: number;
  nonAdmin: number;
  plan: string;
  projects: number;
  restricted: boolean;
  status: string | null;
}

export const BILLING_QUERY = graphql(`
  query Billing {
    billing {
      plan
      status
      currentPeriodEnd
      cancelAtPeriodEnd
      restricted
      canExport
      projects
      admins
      nonAdmin
      maxProjects
      maxAdmins
      maxNonAdmin
      adminEmails
    }
  }
`);

const CREATE_CHECKOUT = graphql(`
  mutation CreateCheckoutSession($interval: BillingInterval!) {
    createCheckoutSession(interval: $interval)
  }
`);

const CREATE_PORTAL = graphql(`
  mutation CreateBillingPortalSession {
    createBillingPortalSession
  }
`);

export const isPaid = (b: Billing | null) => b?.plan === 'crew';

/** A `mailto:` to the org's admins for non-admins to ask about the subscription,
 *  or `null` if there are no admin emails to reach. */
export function contactAdminHref(
  emails: string[],
  subject = 'SiteLens — upgrade to Crew',
): string | null {
  return emails.length > 0
    ? `mailto:${emails.join(',')}?subject=${encodeURIComponent(subject)}`
    : null;
}

/** Fetches the caller org's billing posture. */
export function useBilling() {
  const [billing, setBilling] = useState<Billing | null>(null);
  const [loading, setLoading] = useState(true);

  const reload = useCallback(async () => {
    try {
      const data = await gql(BILLING_QUERY);
      setBilling(data.billing);
    } catch {
      setBilling(null);
    } finally {
      setLoading(false);
    }
  }, []);

  useEffect(() => {
    // eslint-disable-next-line react-hooks/set-state-in-effect
    void reload();
  }, [reload]);

  return { billing, loading, reload };
}

/** Fetches the static plan/feature catalog (drives upgrade UI + selling points). */
export function usePlanCatalog() {
  const [catalog, setCatalog] = useState<PlanCatalog | null>(null);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    let active = true;
    void (async () => {
      try {
        const data = await gql(PLAN_CATALOG_QUERY);
        if (active) {setCatalog(data.planCatalog);}
      } catch {
        if (active) {setCatalog(null);}
      } finally {
        if (active) {setLoading(false);}
      }
    })();
    return () => {
      active = false;
    };
  }, []);

  return { catalog, loading };
}

/** Redirects to hosted Stripe Checkout / the Customer Portal. Admin-only on the
 *  server; on failure we toast and clear `busy` (no redirect happens). */
export function useCheckout() {
  const [busy, setBusy] = useState(false);

  async function startCheckout(interval: 'MONTHLY' | 'ANNUAL') {
    setBusy(true);
    try {
      const { createCheckoutSession } = await gql(CREATE_CHECKOUT, { interval });
      window.location.href = createCheckoutSession;
    } catch (err) {
      toast.error(errMsg(err, 'Could not start checkout'));
      setBusy(false);
    }
  }

  async function openPortal() {
    setBusy(true);
    // Open the tab synchronously (inside the click) so it isn't blocked as a
    // popup after the await; point it at the portal URL once we have it.
    const tab = window.open('', '_blank');
    try {
      const { createBillingPortalSession } = await gql(CREATE_PORTAL);
      if (tab) {
        tab.location.href = createBillingPortalSession;
      } else {
        window.location.href = createBillingPortalSession;
      }
    } catch (err) {
      tab?.close();
      toast.error(errMsg(err, 'Could not open the billing portal'));
    } finally {
      setBusy(false);
    }
  }

  return { busy, openPortal, startCheckout };
}
